src/main/java/org/truffleruby/parser/lexer/RubyLexer.java

```diff
@@ -3665,4 +3665,11 @@ protected boolean isSpaceArg(int c, boolean spaceSeen) {
         return isARG() && spaceSeen && !Character.isWhitespace(c);
     }
 
+    /** Encoding-aware (including multi-byte encodings) check of first character of a given rope, usually to determine
+     * if it is a constant */
+    private boolean isFirstCharacterEncodingAwareUppercase(Rope rope) {
+        byte[] ropeBytes = rope.getBytes();
+        int firstCharacter = rope.encoding.mbcToCode(ropeBytes, 0, ropeBytes.length);
+        return rope.encoding.isUpper(firstCharacter);
+    }
```

Comment:
I'd be curious to see if short-circuiting for ASCII would be beneficial. Something like:

```java
int firstByte = rope.get(0) & 0xFF;
if (rope.encoding.isAsciiCompatible() && isASCII(firstByte)) {
    return StringSupport.isAsciiUppercase(firstByte);
} else {
    final byte[] ropeBytes = rope.getBytes();
    final int firstCodepoint = rope.encoding.mbcToCode(ropeBytes, 0, ropeBytes.length);

    return rope.getEncoding().isUpper(firstCodepoint);
}
```

Depending on the structure of the rope, not having to materialize the whole thing may be advantageous.

It'd be really nice if you could use `RopeNodes.GetCodePointNode` here. But it's probably more expensive than what you're doing here.

Reply:
Regarding the ASCII case, it would be interesting to see if it makes a difference that can be seen in parse time.

Regarding `GetCodePointNode` it would be possible by making it `@GenerateUncached` (not very hard, but seems out of scope of this PR).
I would think most (all?) Ropes in the lexer/parser are SubstringRope.
It could be interesting to have a specialization (or new Rope#getBytesWithOffset(), Rope#getByteOffset() methods) in `GetCodePointNode` to be able to reuse the parent Rope's bytes (+ offset and length) and avoid an extra copy. Although since this is about identifiers they are likely rather short and the copy is likely therefore not expensive.
In the future, maybe TruffleString can support an optimized `getFirstCodePoint()` or so.

Reply:
The ASCII short-circuit seems simple enough, and indeed it'll probably be the most common codepath. I'll update the PR with it.

As for measuring, any suggestions on how to best isolate the parsing?

Reply:
You can use this to measure parse time (either total or per file):
https://github.com/oracle/truffleruby/blob/master/doc/contributor/profiling.md#require-profiling

It's likely much better to run this in the native configuration (`jt build --env native` and then `jt -u native ruby ...`), otherwise it will include some classloading, etc.

Reply:
I've did a few simple tests and the numbers were too close to the margin of error to tell:
    
* `master` (without the fix to #2079): avg 57.7ms, median 57ms, p90 64ms
* with fast path: avg 57.8ms, median 58ms, p90 63ms
* without fast path (previous commit): avg 56.7ms, median 56ms, p90 63ms

Since the fast path is very simple and local to the method, I suggest keeping it anyway.